File: src/mtglab/api/jobs.py
"""In-process job registry for work too slow to answer inside a request.

A Tier 1 sweep at 20,000 games takes around 30 seconds and a land sweep across
eleven counts takes minutes. Neither can run inside a request handler, and
neither is worth a real task queue for a single-user local app -- that would
add a broker, a worker process, and a whole class of "is redis running?"
failures to something a friend is supposed to be able to clone and run.

So: a bounded thread pool plus a dict. Jobs are ephemeral by design. Restart
the server and they are gone, which is correct for a local tool where the
inputs are cheap to resubmit.

**There are two pools, and the split is about what the work waits on.** A Tier 1
sweep is CPU-bound pure Python, so a second simulation thread would contend on
the GIL and make both slower -- one worker, honest queueing. A Claude call is
the opposite shape: it is a socket wait that releases the GIL for minutes at a
time (the theme proposal was measured at 226 seconds), so a single shared queue
would stall a thirty-second sweep behind four minutes of somebody else's
conversation while the CPU sat idle. Hence `CPU` and `NET`. The `NET` lane is
bounded at two for a reason that is not throughput: a proposal costs real money
per run, and a queue is a cheaper way to say "not four at once" than a
rate limiter nobody has written yet.

**Every job carries an owner, and lookups take one.** A job is the first thing
in this app that belongs to a person rather than to the library: its label
names a deck and its result is a simulation somebody paid thirty seconds of CPU
for. `get()` and `all_jobs()` therefore filter rather than trust the caller to,
and a job belonging to somebody else is reported as *absent* — the route turns
`None` into a 404, never a 403, so ids cannot be probed (ADR 5).

The owner is `None` when auth is off, which is every local run: one person,
every job theirs.
"""

from __future__ import annotations

import threading
import traceback
import uuid
from collections.abc import Callable
from concurrent.futures import ThreadPoolExecutor
from dataclasses import dataclass, field
from datetime import UTC, datetime
from typing import Any, Protocol

#: Work that burns CPU: Tier 1, the land sweep. One worker, because the
#: simulation is pure Python and extra threads would contend on the GIL and
#: make every job slower rather than adding throughput.
CPU = "cpu"

#: Work that waits on a socket: anything that calls Anthropic. Two workers,
#: because such a job holds no GIL and blocking a sweep behind it would be a
#: four-minute stall for nothing. See the module docstring for why two.
NET = "net"

#: Work that waits on a Forge subprocess (ADR 35). The thread sleeps in
#: `subprocess.run` while the JVM burns the CPU, so it fits neither lane:
#: in `CPU` it would block Tier 1 sweeps for the minutes a match takes, and
#: in `NET` two matches could run at once — which races the shared `.dck`
#: directory `ensure_profile` hands out and saturates the machine besides.
#: One worker makes both impossible by construction rather than by care.
FORGE = "forge"

_LANES: dict[str, ThreadPoolExecutor] = {
    CPU: ThreadPoolExecutor(max_workers=1, thread_name_prefix="mtglab-job"),
    NET: ThreadPoolExecutor(max_workers=2, thread_name_prefix="mtglab-net"),
    FORGE: ThreadPoolExecutor(max_workers=1, thread_name_prefix="mtglab-forge"),
}
_LOCK = threading.Lock()
_JOBS: dict[str, Job] = {}

# Keep the registry from growing without bound in a long-lived session. The
# bound is global, not per owner, and eviction takes the oldest *finished*
# job — which on a shared instance is very likely one somebody's tab is still
# polling, since a job born `done` (a cache hit) is finished the moment it is
# handed out. Fifty was sized for one laptop; a hundred accounts sharing fifty
# slots would evict results mid-poll under entirely ordinary use. Two hundred
# costs memory only when occupied (a Tier 1 result is a few kB of curves), and
# the LRU-by-`created_at` order is unchanged.
MAX_JOBS = 200

class Progress(Protocol):
    """What a worker is handed to report with: `progress(done, total)`.

    `partial` is the third, optional argument (the match theater): a payload
    the client may render before the result exists — the rows of a match
    still being played. It replaces what was there, is serialised on the job
    while the job runs, and is cleared the moment the job finishes, because
    the result is the whole answer and a leftover partial is a second copy
    that would go stale. Progress was a plain two-argument Callable until
    then; every existing worker still calls it that way, which is why the
    third argument defaults rather than demands.
    """

    # `done` and `total` are positional-only: workers name those parameters
    # freely (`simruns` calls one `_total`), and a Protocol would otherwise
    # bind the names as part of the contract.
    def __call__(self, done: int, total: int, /,
                 partial: Any | None = None) -> None: ...


@dataclass
class Job:
    id: str
    kind: str
    status: str = "queued"        # queued | running | done | error
    done: int = 0
    total: int = 0
    result: Any = None
    # What is known before the result is: the match theater's rows-so-far.
    # Live only while the job is — `submit` clears it on completion.
    partial: Any = None
    error: str | None = None
    label: str = ""
    # Whose job this is. `None` means "the local user", which is everybody when
    # auth is off. Never serialised: a caller who can see a job already knows
    # whose it is, and one who cannot must not learn it exists.
    owner: int | None = None
    #: What this job *is*, for work where asking twice at once is a mistake
    #: rather than a request. `None` opts out and is the default; see `submit`.
    #: Not serialised — it is an internal identity, and for the dossier it is
    #: the cache key, which names a card the caller already knows they asked
    #: about.
    key: str | None = None
    created_at: str = field(
        default_factory=lambda: datetime.now(UTC).isoformat())

    def as_dict(self) -> dict[str, Any]:
        pct = round(100 * self.done / self.total) if self.total else 0
        return {
            "id": self.id,
            "kind": self.kind,
            "status": self.status,
            "done": self.done,
            "total": self.total,
            "percent": pct,
            "partial": self.partial,
            "label": self.label,
            "result": self.result,
            "error": self.error,
            "created_at": self.created_at,
        }


@dataclass
class Plan:
    """What a slow request turns into: an answer, or the work to produce one.

    `result` is set when the answer was already available -- a cached
    simulation, or a Claude mode whose stance is `off` and so made no call. In
    that case `run` is never called. Exactly one of them is used, and the route
    turns that into a job born finished or a job that was queued.

    It lives here rather than beside either caller because it is the shape of a
    job's *input*, and there are now two producers of one: `api/simruns.py` and
    `api/themeruns.py`. `lane` is the second reason -- which pool the work
    belongs in is a property of the work, so it is decided where the work is
    planned rather than at the route, where somebody would eventually forget.
    """

    kind: str
    label: str
    result: dict[str, Any] | None
    run: Callable[[Progress], dict[str, Any]]
    lane: str = CPU
    #: What this work *is*, when two simultaneous requests for it are the same
    #: request. Decided by the planner because only the planner knows what
    #: identity means for its own work — for the dossier it is the commander,
    #: not the deck. `None` means "start a second one", which is right for
    #: anything not reproducible; see `submit`.
    key: str | None = None


#: A job that has not finished. Two of these for the same `key` is the bug
#: `submit` exists to prevent; a finished one is a cache's problem, not this
#: module's.
LIVE = ("queued", "running")


def _keep_locked() -> None:
    """Evict finished jobs if the registry is over its bound. `_LOCK` held."""
    if len(_JOBS) <= MAX_JOBS:
        return
    finished = [j for j in _JOBS.values() if j.status not in LIVE]
    for old in sorted(finished, key=lambda j: j.created_at)[:len(_JOBS) - MAX_JOBS]:
        _JOBS.pop(old.id, None)


def _record(job: Job) -> Job:
    """Register a job, evicting finished ones if the registry is over its bound."""
    with _LOCK:
        _JOBS[job.id] = job
        _keep_locked()
    return job


def completed(kind: str, *, result: Any, label: str = "",
              owner: int | None = None) -> Job:
    """A job that is already finished, because its answer was already known.

    This is what a simulation cache hit returns (`sim/cache.py`). The
    alternative was a second response shape saying "no job, here is the
    result", and it was rejected: every client would need a branch, a hit would
    leave no trace in `/api/jobs`, and the claim being made is not "this is not
    a job" but "this job took no time". A job born done is the honest shape and
    the response contract does not fork.

    It still costs a registry slot, which is the point of `_record` being
    shared -- fifty instant hits evict fifty finished jobs, not the running one.
    """
    return _record(Job(id=uuid.uuid4().hex[:12], kind=kind, status="done",
                       done=1, total=1, result=result, label=label,
                       owner=owner))


def submit(kind: str, fn: Callable[[Progress], Any],
           *, label: str = "", owner: int | None = None,
           lane: str = CPU, key: str | None = None) -> Job:
    """Queue `fn`, handing it a `progress(done, total)` callback to report with.

    `lane` picks the pool -- `CPU` for work that computes, `NET` for work that
    waits on Anthropic. Checked before the job is recorded, so a typo is an
    exception out of the route rather than a job that sits `queued` forever
    because nothing was ever going to run it.

    **`key` makes asking twice at once return the same job, and that is a
    money bug rather than a tidiness one.** A dossier takes about four minutes
    and pays for a web search; nothing stopped a second click, a second tab or
    a second device inside that window from starting a *second* paid run for
    the same commander, and on 2026-08-13 two ran concurrently on the deployed
    instance for exactly that reason. The client-side reattach that was built
    first only ever covered one tab, because it lives in that tab's
    localStorage; this covers every case at once by making the server the
    thing that knows a run is already going.

    Matching is per **owner** as well as per key, which is deliberate rather
    than cautious: a job belongs to a person (ADR 5) and `get` refuses one that
    is not yours, so handing two accounts the same id would give the second a
    404 for a job it had just been told about. Every case the bug actually
    covers -- one person, two tabs -- is one owner.

    Only a *live* job is reused. A finished one is what a cache is for, and the
    dossier has one (ADR 19); a failed one must be retryable or a transient
    Anthropic error would be permanent until the process restarted.

    The lookup and the insert are one locked step, because the window this
    closes is exactly the window two concurrent requests race in.
    """
    pool = _LANES.get(lane)
    if pool is None:
        raise ValueError(
            f"unknown job lane {lane!r}; expected one of {sorted(_LANES)}")

    with _LOCK:
        if key:
            already = next(
                (j for j in _JOBS.values()
                 if j.key == key and j.kind == kind and j.owner == owner
                 and j.status in LIVE),
                None)
            if already is not None:
                return already
        job = Job(id=uuid.uuid4().hex[:12], kind=kind, label=label,
                  owner=owner, key=key)
        _JOBS[job.id] = job
        _keep_locked()

    def progress(done: int, total: int, partial: Any | None = None) -> None:
        job.done, job.total = done, total
        if partial is not None:
            job.partial = partial

    def wrapped() -> None:
        job.status = "running"
        try:
            job.result = fn(progress)
            job.status = "done"
            if job.total:
                job.done = job.total
            # The result is the whole answer; a partial left behind is a
            # stale second copy of part of it.
            job.partial = None
        except Exception as exc:                                    # noqa: BLE001
            # Surface the message and NOT the exception's class name.
            #
            # This read `f"{type(exc).__name__}: {exc}"` until 2026-08-22, so a
            # failed job rendered `DeckNotFound: no deck 'nope'` in the
            # browser -- `job.error` becomes a JS `Error` in `lib/api.ts` and
            # the screen shows it. A Python class name is a technology a user
            # can see, which commandment 10 does not allow; Aaron ruled on it
            # when the Go port surfaced the seam, since `internal/jobs`
            # records `err.Error()` and has no class name to offer. Correcting
            # Python rather than teaching Go to imitate it is what makes the
            # two agree *now*, instead of at whichever flip moves this family.
            #
            # Nothing is lost. Every exception a job worker can raise carries
            # a message -- checked, there are no bare raises -- so the prefix
            # was never the load-bearing half, and `traceback.print_exc()`
            # still puts the class name in the log, where a maintainer reads
            # it and a visitor does not.
            job.error = str(exc)
            job.status = "error"
            traceback.print_exc()

    pool.submit(wrapped)
    return job


def get(job_id: str, *, owner: int | None = None) -> Job | None:
    """One job, if it is this owner's. `None` covers both "no such job" and
    "not yours", which is the point -- the caller cannot tell them apart, and
    neither can whoever is probing ids."""
    with _LOCK:
        job = _JOBS.get(job_id)
    return job if job is not None and job.owner == owner else None


def all_jobs(*, owner: int | None = None) -> list[Job]:
    """This owner's jobs, newest first."""
    with _LOCK:
        mine = [j for j in _JOBS.values() if j.owner == owner]
    return sorted(mine, key=lambda j: j.created_at, reverse=True)


def census() -> dict[str, int]:
    """How many jobs the registry holds, by status. Counts and nothing else.

    The admin dashboard's view of this registry, and deliberately the only
    cross-owner one: `get` and `all_jobs` scope to an owner because a job's
    label can name another person's deck, and administering the instance
    (ADR 17) is about load, not about reading anybody's work. Counts cannot
    carry a name.
    """
    with _LOCK:
        counts: dict[str, int] = {}
        for job in _JOBS.values():
            counts[job.status] = counts.get(job.status, 0) + 1
    return counts


def forget_owner(owner: int) -> int:
    """Drop every job belonging to one account. Returns how many went.

    Called when an account is deleted, and the reason is `users.id`: it is
    `INTEGER PRIMARY KEY` without `AUTOINCREMENT`, so SQLite is free to re-issue
    a deleted account's rowid to the next account created. Jobs are keyed on
    that integer and held in memory, so without this the next holder of the id
    would inherit the dead account's jobs -- results, and the deck names in
    their labels. That is the isolation `get` and `all_jobs` are written to
    enforce, defeated by arithmetic rather than by a missing filter.

    A running job is dropped from the registry but not cancelled; its thread
    finishes into a record nothing points at. That is the honest trade -- the
    pools have no cancellation and inventing one here would be a worse lie than
    a few seconds of orphaned CPU.

    `owner` is never `None` here: that is the no-auth local case, where there is
    one person and no account to delete.
    """
    with _LOCK:
        doomed = [k for k, j in _JOBS.items() if j.owner == owner]
        for key in doomed:
            del _JOBS[key]
    return len(doomed)


def clear() -> None:
    """Test helper -- drops every recorded job."""
    with _LOCK:
        _JOBS.clear()

File: go/internal/jobs/jobs.go
// Package jobs is the in-process registry for work
// too slow to answer inside a request.
//
// The argument for a registry in one process -- rather than a broker, a
// worker process and a whole class of "is redis running?" failures -- comes
// with a consequence embraced on purpose: **jobs are ephemeral by
// design.** Restart the process and they are gone, which is correct for
// inputs that are cheap to resubmit.
//
// # The CPU lane follows the machine
//
// The CPU lane is a
// **semaphore over goroutines**, sized from the machine (see [Lane] and
// [Config.CPUWorkers]), so two sweeps genuinely run at once on a machine with
// two cores to run them on.
//
// **The instance has one core, so the dividend is banked and not yet
// realised.** `nproc` on the deployed machine answers 1 (measured
// 2026-08-22, and `shared-cpu-1x` is a 1-vCPU microVM rather than a quota),
// which makes the CPU lane today one-at-a-time queueing. ADR 38 named
// the width as something worth buying, and what is bought is the
// *capability* -- the
// width follows the machine, so scaling the instance widens the lane with no
// code change and no second process to keep a registry in step. Until then,
// quoting it as a throughput win would be quoting a measurement nobody has
// taken, and N-core Tier 1 scaling cannot be measured on
// `shared-cpu-1x` at all.
//
// Three things that widening it changes downstream, written here because the
// next session will meet them and not here:
//
//   - **The sim family has no dedupe key.** The sim routes pass none: they
//     lean on ADR 18's cache instead, which covers "somebody asked before"
//     but not "somebody is asking right now". Two identical sweeps submitted
//     together compute side by side and both store.
//     That is a wasted core, never a wrong answer -- and `sim/cache` opens
//     `app.db` in WAL with a 5s busy timeout and never fails the caller, so
//     a collision
//     is a short wait and, at worst, a logged warning.
//   - **Completion order is not submission order.** Nothing on the wire
//     promises it; `GET /api/jobs` sorts by `created_at`, which is a
//     submission stamp and unaffected.
//   - **A queued job is a parked goroutine.** Roughly 4KB of stack each, and
//     the
//     live count is bounded by nothing but the request rate ([MaxJobs] evicts
//     only *finished* jobs). At the concurrency `fly.toml` admits -- 40 hard
//     -- that is under a megabyte, so the simpler structure wins; a bounded
//     queue would have to block the submitting request, which the
//     unbounded one never does.
//
// # What is fixed regardless of the machine
//
// **The lanes are still three, and two of them are not about the machine.**
// `NET` stays at two because a Claude call costs real money per run and a
// queue is a cheaper way to say "not four at once" than a rate limiter
// nobody has written; `FORGE` stays at one because two JVMs would race the
// shared `.dck` directory `ensure_profile` hands out. Sizing either from
// `GOMAXPROCS` would be reading the wrong fact off the wrong machine.
//
// **Every job carries an owner, and lookups take one.** A job is the first
// thing in this app that belongs to a person rather than to the library: its
// label names a deck and its result is a simulation somebody paid for. [Registry.Get]
// and [Registry.All] filter rather than trust the caller, and a job belonging
// to somebody else is reported as *absent* -- the route turns that into a
// 404, never a 403, so ids cannot be probed (ADR 5).
//
// **There is no cancellation, and inventing one here would be a worse lie
// than a few seconds of orphaned CPU.** A dropped job's goroutine finishes
// into a record nothing
// points at. No `context.Context` reaches a worker for that reason -- a
// context that is never cancelled is a promise the caller would believe.
//
// # Two guards the concurrency demands
//
// **The mutable half of a [Job] is guarded.** A worker writes a job's
// status, progress and result while request goroutines read them for polls,
// and the race
// detector makes an unguarded version a hard failure. So the fields a worker
// touches live
// behind the registry's mutex and [Job.Payload] takes it. A poll sees a
// torn-free snapshot of one moment.
//
// **A panicking worker is a failed job, not a dead process.** A bug in one
// job must cost one job's `error`
// string. Go's default is the opposite -- an unrecovered panic in any
// goroutine takes the process with it -- and for a door that is also serving
// every other request, that would be strictly worse.
// The runner recovers, records it as `panic: <what panicked>` so a reader can
// tell a crash from a refusal the worker made on purpose, and logs the
// stack.
//
// # A recorded seam, still open
//
// The recorded failure string is a kind name and a message, e.g.
// `DeckNotFound: no deck 'nope'` -- and that string
// reaches the browser. A bare `err.Error()` has
// no kind name in it. **Whether a job family reproduces the prefix is
// that family's decision, taken at its own site**, not something this
// package
// can decide for them; it is recorded here because the alternative is that
// nobody notices the strings changed. (There is a second question underneath
// it, for Aaron: an internal error-kind name on a user-visible surface sits
// oddly beside commandment 10.)
package jobs

import (
	"fmt"
	"math"
	"strings"
	"time"
)

// Lane is which pool work belongs in, and it is a property of the work rather
// than of the route -- which is why a [Plan] carries one and a handler never
// picks.
type Lane string

const (
	// CPU is work that burns CPU: Tier 1, the land sweep, the mulligan grid.
	// Wide as the machine allows.
	CPU Lane = "cpu"

	// NET is work that waits on a socket: anything that calls Anthropic. Two
	// at once, and the bound is about money rather than throughput -- see the
	// package comment.
	NET Lane = "net"

	// FORGE is work that waits on a Forge subprocess (ADR 35). It fits
	// neither of the others: in CPU it would block Tier 1 for the minutes a
	// match takes, and at NET's width two matches could run at once, which
	// races the shared `.dck` directory and saturates the machine besides.
	// One worker makes both impossible by construction rather than by care.
	FORGE Lane = "forge"
)

// Lanes are the three, in the order the served error message lists them --
// alphabetical, which happens to read cpu, forge,
// net.
var Lanes = []Lane{CPU, FORGE, NET}

// Status is where a job is. The four strings are wire contract, because
// `web/src/lib/api.ts` compares against them.
const (
	Queued  = "queued"
	Running = "running"
	Done    = "done"
	Errored = "error"
)

// live reports whether a status is one a second ask may join. A finished job
// is a cache's problem, not this module's, and a failed one must stay
// retryable or a transient Anthropic error would be permanent until the
// process restarted.
func live(status string) bool { return status == Queued || status == Running }

// MaxJobs bounds the registry, so a long-lived session does not grow without
// one. The bound is global rather than per owner, and eviction takes the
// oldest *finished* job -- which on a shared instance is very likely one
// somebody's tab is still polling, since a job born `done` (a cache hit) is
// finished the moment it is handed out. Fifty was sized for one laptop; two
// hundred costs memory only when occupied.
const MaxJobs = 200

// Progress is what a worker is handed to report with.
//
// Two methods rather than one variadic call, because an omitted partial
// means something different from a passed one:
// omitted leaves whatever was there alone, and only a real payload
// replaces it. An interface says that where a `nil` would have to be
// explained.
//
// Partial is the match theater's: rows of a match still being played,
// renderable before the result exists. It is serialised on the job while the
// job runs and **cleared the moment the job finishes**, because the result is
// the whole answer and a leftover partial is a second copy that would go
// stale.
type Progress interface {
	// Report is the two-argument call every worker
	// written before the theater still makes.
	Report(done, total int)
	// ReportPartial also replaces the partial payload.
	ReportPartial(done, total int, partial any)
}

// Runner is the work itself: handed a Progress, it returns the answer or
// the error.
type Runner func(Progress) (any, error)

// Plan is what a slow request turns into: an answer, or the work to produce
// one.
//
// Result is set when the answer was already available -- a cached simulation,
// or a Claude mode whose stance is `off` and so made no call. In that case
// Run is never called. Exactly one of them is used, and [Registry.FromPlan]
// turns that into a job born finished or a job that was queued.
//
// It lives here rather than beside any caller because it is the shape of a
// job's *input*, and there are several producers of one. Lane is the second
// reason: which pool the work belongs in is a property of the work, so it is
// decided where the work is planned rather than at the route, where somebody
// would eventually forget.
type Plan struct {
	Kind  string
	Label string
	// Result short-circuits: non-nil means the answer is already known.
	Result any
	Run    Runner
	// Lane defaults to CPU when empty.
	Lane Lane
	// Key is what this work *is*, when two simultaneous requests for it are
	// the same request. Decided by the planner, because only the planner
	// knows what identity means for its own work -- for the dossier it is the
	// commander, not the deck. Empty means "start a second one", which is
	// right for anything not reproducible; see [Registry.Submit].
	Key string
}

// UnknownLaneError is what [Registry.Submit] answers for a lane nothing will
// ever run, checked before the job is recorded so a typo is an error out of
// the route rather than a job that sits `queued` forever.
//
// Its message is pinned byte for byte, quoting style and bracketed list
// included, because that string is what a 500 carries and the recorded
// golden holds it.
type UnknownLaneError struct{ Lane Lane }

func (e *UnknownLaneError) Error() string {
	names := make([]string, 0, len(Lanes))
	for _, lane := range Lanes {
		names = append(names, quoted(string(lane)))
	}
	return fmt.Sprintf("unknown job lane %s; expected one of [%s]",
		quoted(string(e.Lane)), strings.Join(names, ", "))
}

// quoted renders a string the way the served message has always quoted
// one, which is not `strconv.Quote`: **single
// quotes preferred**, switching to double ones only when the string
// holds a single quote and no double. That is the whole of the rule for
// anything a lane can be -- a short identifier out of a planner -- and the
// corpus carries a quoted case in each direction so the preference is
// checked rather than assumed. Escapes beyond the quote, the backslash and
// the three whitespace ones are out of reach here and are not reproduced.
func quoted(s string) string {
	quote := byte('\'')
	if strings.Contains(s, "'") && !strings.Contains(s, `"`) {
		quote = '"'
	}
	var out strings.Builder
	out.WriteByte(quote)
	for _, r := range s {
		switch r {
		case rune(quote):
			out.WriteByte('\\')
			out.WriteByte(quote)
		case '\\':
			out.WriteString(`\\`)
		case '\n':
			out.WriteString(`\n`)
		case '\r':
			out.WriteString(`\r`)
		case '\t':
			out.WriteString(`\t`)
		default:
			out.WriteRune(r)
		}
	}
	out.WriteByte(quote)
	return out.String()
}

// Payload is what a poll gets back. The field order is
// the recorded wire order, so the bytes match as well as the keys.
//
// Owner and Key are deliberately absent. Owner because a caller who can see a
// job already knows whose it is, and one who cannot must not learn it exists;
// Key because it is an internal identity -- and for the dossier it is the
// cache key, which names a card.
type Payload struct {
	ID        string  `json:"id"`
	Kind      string  `json:"kind"`
	Status    string  `json:"status"`
	Done      int     `json:"done"`
	Total     int     `json:"total"`
	Percent   int     `json:"percent"`
	Partial   any     `json:"partial"`
	Label     string  `json:"label"`
	Result    any     `json:"result"`
	Error     *string `json:"error"`
	CreatedAt string  `json:"created_at"`
}

// percent is the progress percentage, zero when the total is, and the only
// interesting word in its arithmetic is `round`.
//
// **The recorded rounding is half to even, and Go's `math.Round` rounds half
// away from
// zero**, so the two disagree one case in eight: at done=1, total=8 the true
// value is
// 12.5, and the corpus answers 12 where `math.Round` answers 13. The corpus
// in
// `testdata/jobs.json` is what caught it, and it is checked rather than
// argued because a one-point difference in a progress bar is exactly the kind
// of wrongness nobody reports.
//
// The arithmetic is done in float64 for the same reason: the recorded value
// is a true division, so the value being rounded is a
// float with a float's representation, and integer arithmetic here would
// disagree on its own schedule.
func percent(done, total int) int {
	if total == 0 {
		return 0
	}
	return roundHalfToEven(float64(100*done) / float64(total))
}

// roundHalfToEven is the nearest integer, ties to the even one: a floor,
// moved up when the fraction exceeds a half, and on an exact half only when
// the floor is odd.
func roundHalfToEven(x float64) int {
	nearest := math.Floor(x)
	frac := x - nearest
	switch {
	case frac > 0.5:
		nearest++
	case frac == 0.5 && math.Mod(nearest, 2) != 0:
		nearest++
	}
	return int(nearest)
}

// stamp is the app's recorded timestamp format.
//
// Two details, and both are load-bearing rather than cosmetic. **The
// fractional part vanishes entirely when there is none** -- the format
// writes
// `...:20+00:00`, never `...:20.000000+00:00` -- which happens once in a
// million and would otherwise be a shape no client had seen. And **the offset
// is spelled `+00:00`, never `Z`**, which is what `web/src/lib/api.ts` has
// always been handed.
//
// Both matter more than they look, because this string is also the **sort
// key**: the job listing sorts on the text, not on an instant. That is
// sound -- every field is fixed width, and within one second the elided form
// begins with `+` (0x2B), which sorts below every digit the six-digit form
// can start with, so the microsecond-zero job correctly comes first. Storing
// the string rather than the instant is therefore the sound choice, not a
// shortcut.
func stamp(t time.Time) string {
	t = t.UTC().Truncate(time.Microsecond)
	if t.Nanosecond() == 0 {
		return t.Format("2006-01-02T15:04:05") + "+00:00"
	}
	return t.Format("2006-01-02T15:04:05.000000") + "+00:00"
}
